modules/core/src/directx.cpp

```diff
@@ -236,187 +237,121 @@ int getTypeFromD3DFORMAT(const int iD3DFORMAT)
-
-    cl_platform_id getPlatform()
+class OpenCL_D3D11_NV : public ocl::Context::UserContext
+{
+public:
+    OpenCL_D3D11_NV(cl_platform_id platform, ID3D11Device*_device) : device(_device)
     {
-        if (!platform_)
+        clCreateFromD3D11Texture2DNV = (clCreateFromD3D11Texture2DNV_fn)
+            clGetExtensionFunctionAddressForPlatform(platform, "clCreateFromD3D11Texture2DNV");
+        clEnqueueAcquireD3D11ObjectsNV = (clEnqueueAcquireD3D11ObjectsNV_fn)
+            clGetExtensionFunctionAddressForPlatform(platform, "clEnqueueAcquireD3D11ObjectsNV");
+        clEnqueueReleaseD3D11ObjectsNV = (clEnqueueReleaseD3D11ObjectsNV_fn)
+            clGetExtensionFunctionAddressForPlatform(platform, "clEnqueueReleaseD3D11ObjectsNV");
+        if (!clCreateFromD3D11Texture2DNV || !clEnqueueAcquireD3D11ObjectsNV || !clEnqueueReleaseD3D11ObjectsNV)
         {
-            CV_Assert(cv::ocl::haveOpenCL());
-
-            cl_device_id device = (cl_device_id)ocl::Device::getDefault().ptr();
-            CV_Assert(device);
-            cl_int status = clGetDeviceInfo(device, CL_DEVICE_PLATFORM, sizeof(platform_), &platform_, NULL);
-            if (status != CL_SUCCESS)
-                CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get platform corresponding to device");
+            CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't find functions for D3D11_NV");
         }
-
-        return platform_;
+        device->AddRef();
     }
-
-
-    bool initializeD3D11()
-    {
-        using namespace cv::ocl;
-        cl_platform_id platform = getPlatform();
-
-        bool useCLNVEXT = false;
-        size_t exts_len;
-        cl_int status = clGetPlatformInfo(platform, CL_PLATFORM_EXTENSIONS, 0, NULL, &exts_len);
-        if (status != CL_SUCCESS)
-            CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't get length of CL_PLATFORM_EXTENSIONS");
-        cv::AutoBuffer<char> extensions(exts_len);
-        status = clGetPlatformInfo(platform, CL_PLATFORM_EXTENSIONS, exts_len, static_cast<void*>(extensions.data()), NULL);
-        if (status != CL_SUCCESS)
-            CV_Error(cv::Error::OpenCLInitError, "OpenCL: No available CL_PLATFORM_EXTENSIONS");
-        bool is_support_cl_khr_d3d11_sharing = false;
-        if (strstr(extensions.data(), "cl_khr_d3d11_sharing"))
-            is_support_cl_khr_d3d11_sharing = true;
-#ifdef HAVE_OPENCL_D3D11_NV
-        bool is_support_cl_nv_d3d11_sharing = false;
-        if (strstr(extensions.data(), "cl_nv_d3d11_sharing"))
-            is_support_cl_nv_d3d11_sharing = true;
-        if (!is_support_cl_nv_d3d11_sharing && !is_support_cl_khr_d3d11_sharing)
-            CV_Error(cv::Error::OpenCLInitError, "OpenCL: No supported extensions");
-#else
-        if (!is_support_cl_khr_d3d11_sharing)
-            CV_Error(cv::Error::OpenCLInitError, "OpenCL: No supported extensions");
+    ~OpenCL_D3D11_NV() {
+        device->Release();
```

Comment:
Can't reproduce in my environment so far. Could be (at least) two potential reasons
1. Clean-up order - as it's after ExitProcess, is it possible that some global d3d11/dxgi resources already closed at time device Release called?
2. Incorrect manipulation (directly or indirectly) with d3d device reference counter, so that Release called on already destroyed object

I tried to check that it's not 2 (which is bad) by inserting
```
ocl::OpenCLExecutionContext::getCurrent().release()
```
into sample `video_acceleration` right before exiting function main, works ok.
For similar experiment in tests, is there easy way to close OpenCL context after all tests finished but before ExitProcess?

Reply:
> is there easy way to close OpenCL context after all tests finished

We can do that in our tests. But problem itself would migrate to OpenCV Users because we just "hides" that.

---

We have similar issues with OpenCL, so there are several workarounds to avoid calling of any release functions after "ExitProcess" (see `__termination` flag).

Reply:
I mean modify tests temporary for experimental purpose - if call to release OpenCL context before ExitProcess causes crash, it definitely requires investigation and fix; if no crash, should be ok to add workarounds targeted specifically for termination process?

Reply:
> for experimental purpose

Adding `ocl::OpenCLExecutionContext::getCurrent().release();` into the test's main() function eliminates crash.

---

`__termination` flag is not enabled in this problematic code flow (`DllMain()` is not called, cleanup is performed through FLS):

```
...
20 000000e6`622ff790 00007ff9`ddc6b690     opencv_core452!cv::TLSData<cv::CoreTLSData>::deleteDataInstance+0x47 [C:\build\precommit_custom_windows\opencv\modules\core\include\opencv2\core\utils\tls.hpp @ 82] 
21 000000e6`622ff7c0 00007ffa`18eeddd5     opencv_core452!cv::details::TlsStorage::releaseThread+0x130 [C:\build\precommit_custom_windows\opencv\modules\core\src\system.cpp @ 1571] 
22 000000e6`622ff800 00007ffa`18eedae8     ntdll!RtlpFlsDataCleanup+0x121
23 000000e6`622ff840 00007ffa`18eeda2d     ntdll!LdrShutdownProcess+0x98
24 000000e6`622ff940 00007ffa`1731e0ab     ntdll!RtlExitUserProcess+0xad
25 000000e6`622ff970 00007ffa`16a405bc     KERNEL32!ExitProcessImplementation+0xb
26 000000e6`622ff9a0 00007ffa`16a4045f     ucrtbase!exit_or_terminate_process+0x44
27 000000e6`622ff9d0 00007ff7`81b0718b     ucrtbase!common_exit+0x6f
28 000000e6`622ffa20 00007ffa`17317034     opencv_test_videoio!__scrt_common_main_seh+0x173 [D:\agent\_work\9\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 295] 
29 000000e6`622ffa60 00007ffa`18ee2651     KERNEL32!BaseThreadInitThunk+0x14
2a 000000e6`622ffa90 00000000`00000000     ntdll!RtlUserThreadStart+0x21
```

Will take a look how to deal this that case.